armi/reactor/converters/axialExpansionChanger/axialExpansionChanger.py

```diff
@@ -386,6 +443,134 @@ def axiallyExpandAssembly(self):
+        ----------
+        fromComp
+            Component which is going to give mass to toComp
+        toComp
+            Component that is recieving mass from fromComp
+        deltaZTop
+            The length, in cm, of fromComp being given to toComp
+
+        Notes
+        -----
+        Only the mass of ``toComp`` is changed in this method. The mass of ``fromComp`` is changed separately by
+        changing the height of ``fromComp`` -- the number densities of ``fromComp`` are not modified. When
+        redistributing mass, if ``fromComp`` and ``toComp`` are different temperatures, the temperature of
+        ``toComp`` will change. Calculating this new temperature is non trivial due to thermal expansion. The following
+        defines what the area of ``toComp`` is post-redistribution,
+
+        .. math::
+
+            A_1(\hat{T}) \left( H_1 + \delta \right) &= A_1(T_1) H_1 + A_2(T_2)\delta,
+            A_1(\hat{T}) &= \frac{A_1(T_1) H_1 + A_2(T_2)\delta}{H_1 + \delta}.
+
+        Where, :math`A_1, T_1, H_1`, are the area, temperature, and height of ``toComp``, :math:`A_2, T_2`, are the
+        area and temparature of ``fromComp``, :math:`\delta` is the parameter ``deltaZTop``, and :math:`\hat{T}` is
+        the new temperature of ``toComp`` post-redistribution. :func:`scipy.optimize.brentq` is used to
+        find the root of the above equation, indicating the value for :math:`\hat{T}`
+        that finds the desired area, post-redistribution of mass.
+
+        Raises
+        ------
+        RuntimeError if the linked components are not the same material; we cannot transfer mass between materials
+        because then the resulting material has unknown properties.
+        """
+        # limitation: fromComp and toComp **must** be the same materials.
+        if type(fromComp.material) is not type(toComp.material):
+            msg = f"""
+            Cannot redistribute mass between components that are different materials!
+                Trying to redistribte mass between the following components in {self.linked.a}:
+                    from --> {fromComp.parent} : {fromComp} : {type(fromComp.material)}
+                      to --> {toComp.parent} : {toComp} : {type(toComp.material)}
+
+                Instead, mass will be removed from ({fromComp} | {type(fromComp.material)}) and
+                ({toComp} | {type(toComp.material)} will be artificially expanded. The consequence is that mass
+                conservation is no longer guaranteed for the {toComp.getType()} component type on this assembly!
+            """
+            runLog.warning(dedent(msg))
+            return
+
+        toCompVolume = toComp.getArea() * toComp.height
+        fromCompVolume = fromComp.getArea() * abs(deltaZTop)
+        newVolume = fromCompVolume + toCompVolume
+
+        ## calculate the mass of each nuclide and then the ndens for the new mass
+        newNDens: dict[str, float] = {}
+        for nuc in fromComp.getNuclides():
```

Comment:
```suggestion
        allNuclides = set(fromComp.getNuclides()) + set(toComp.getNuclides())
        for nuc in allNuclides:
```

If `toComp` has any nuclides that are not in `fromComp`, they will get dropped here. It's probably not that common because we're forcing the two components to be the same material, but it seems more robust to check for all nuclides.

Reply:
https://github.com/terrapower/armi/pull/2198/commits/18b0596eeae275e28c0311d7b56a81c8dd897d61